File: backend/src/models/Perhitungan.js
const mongoose = require('mongoose');

const perhitunganSchema = new mongoose.Schema({
    userId: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'User',
        required: true,
        index: true
    },
    idDeteksi: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'DeteksiYOLO',
        sparse: true // Make it optional for manual calculations
    },
    idVideo: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'Video',
        sparse: true,
        index: true
    },
    waktuProses: {
        type: Date,
        default: Date.now
    },
    // Results
    jumlahMobil: {
        type: Number,
        required: true,
        min: 0,
        default: 0
    },
    jumlahMotor: {
        type: Number,
        required: true,
        min: 0,
        default: 0
    },
    totalKendaraan: {
        type: Number,
        required: true,
        min: 0,
        default: 0
    },
    // Calculations
    DJ: {
        type: Number,
        required: true,
        min: 0
    },
    LOS: {
        type: String,
        enum: ['A', 'B', 'C', 'D', 'E', 'F'],
        required: true
    },
    // Additional metrics
    metrics: {
        namaRuas: String,
        tipeJalan: String,
        kapasitas: Number,
        flowRate: Number, // Volume in SMP/jam
        durasiMenit: Number,
        waktuObservasi: String,
        deteksiId: mongoose.Schema.Types.ObjectId,
        averageSpeed: Number, // km/h (optional)
        density: Number, // vehicles per km (optional)
        peakHour: {
            start: String,
            end: String,
            count: Number
        }
    }
}, {
    timestamps: true
});

// Indexes
perhitunganSchema.index({ idVideo: 1 });
perhitunganSchema.index({ waktuProses: -1 });

// Calculate total before saving
perhitunganSchema.pre('save', function(next) {
    this.totalKendaraan = this.jumlahMobil + this.jumlahMotor;
    next();
});

// Static method to calculate DJ (Derajat Jenuh)
perhitunganSchema.statics.calculateDJ = function(jumlahMobil, jumlahMotor) {
    // Formula DJ = (Q / C)
    // Q = Volume lalu lintas (smp/jam)
    // C = Kapasitas jalan (smp/jam)
    
    // Konversi ke satuan mobil penumpang (SMP)
    // 1 mobil = 1 SMP
    // 1 motor = 0.5 SMP
    const totalSMP = jumlahMobil + (jumlahMotor * 0.5);
    
    // Asumsi kapasitas jalan (bisa disesuaikan)
    const kapasitasJalan = 2000; // SMP per jam
    
    const DJ = totalSMP / kapasitasJalan;
    return Math.min(DJ, 1); // Max 1
};

// Static method to calculate LOS based on DJ
perhitunganSchema.statics.calculateLOS = function(DJ) {
    if (DJ <= 0.20) return 'A';
    if (DJ <= 0.44) return 'B';
    if (DJ <= 0.72) return 'C';
    if (DJ <= 0.84) return 'D';
    if (DJ <= 0.92) return 'E';
    return 'F';
};

// Method to get full calculation result
perhitunganSchema.methods.getFullResult = function() {
    return {
        perhitunganId: this._id,
        videoId: this.idVideo,
        deteksiId: this.idDeteksi,
        waktuProses: this.waktuProses,
        hasil: {
            jumlahMobil: this.jumlahMobil,
            jumlahMotor: this.jumlahMotor,
            totalKendaraan: this.totalKendaraan,
            DJ: this.DJ,
            LOS: this.LOS
        },
        metrics: this.metrics
    };
};

// Virtual for LOS description
perhitunganSchema.virtual('losDescription').get(function() {
    const descriptions = {
        'A': 'Kondisi arus bebas dengan kecepatan tinggi',
        'B': 'Arus stabil dengan kecepatan mulai dibatasi',
        'C': 'Arus stabil dengan kecepatan terbatas',
        'D': 'Arus mendekati tidak stabil',
        'E': 'Arus tidak stabil, sering berhenti',
        'F': 'Arus dipaksa atau macet'
    };
    return descriptions[this.LOS] || 'Tidak diketahui';
});

module.exports = mongoose.model('Perhitungan', perhitunganSchema);

File: backend/app/models/perhitungan.py
"""
Perhitungan Model - Pydantic schemas for Traffic Calculations
"""

from datetime import datetime
from typing import Optional, Dict, Any
from pydantic import BaseModel, Field


class PerhitunganMetrics(BaseModel):
    namaRuas: Optional[str] = None
    tipeJalan: Optional[str] = None
    kapasitas: Optional[int] = None
    flowRate: Optional[float] = None
    durasiMenit: Optional[int] = None
    waktuObservasi: Optional[str] = None
    deteksiId: Optional[str] = None
    averageSpeed: Optional[float] = None
    density: Optional[float] = None
    peakHour: Optional[Dict[str, Any]] = None


class PerhitunganBase(BaseModel):
    jumlahMobil: int = Field(default=0, ge=0)
    jumlahMotor: int = Field(default=0, ge=0)
    totalKendaraan: int = Field(default=0, ge=0)
    DJ: float = Field(..., ge=0)
    LOS: str = Field(..., pattern="^[A-F]$")


class PerhitunganCreate(PerhitunganBase):
    userId: str
    idDeteksi: Optional[str] = None
    idVideo: Optional[str] = None
    metrics: Optional[PerhitunganMetrics] = None


class PerhitunganUpdate(BaseModel):
    jumlahMobil: Optional[int] = Field(None, ge=0)
    jumlahMotor: Optional[int] = Field(None, ge=0)
    DJ: Optional[float] = Field(None, ge=0)
    LOS: Optional[str] = Field(None, pattern="^[A-F]$")
    metrics: Optional[PerhitunganMetrics] = None


class PerhitunganInDB(PerhitunganBase):
    id: str = Field(alias="_id")
    userId: str
    idDeteksi: Optional[str] = None
    idVideo: Optional[str] = None
    waktuProses: datetime = Field(default_factory=datetime.utcnow)
    metrics: Optional[PerhitunganMetrics] = None
    createdAt: datetime = Field(default_factory=datetime.utcnow)
    updatedAt: datetime = Field(default_factory=datetime.utcnow)

    class Config:
        populate_by_name = True


class PerhitunganResponse(BaseModel):
    id: str
    userId: str
    idDeteksi: Optional[str] = None
    jumlahMobil: int
    jumlahMotor: int
    totalKendaraan: int
    DJ: float
    LOS: str
    metrics: Optional[PerhitunganMetrics] = None
    createdAt: Optional[datetime] = None

    class Config:
        populate_by_name = True


class ManualCalculationRequest(BaseModel):
    # Data jalan
    namaRuas: str
    tipeJalan: str = "4/2 D"
    jumlahLajur: int = 2
    lebarLajur: float = 3.5
    faktorPemisah: str = "50-50"
    hambatanSamping: str = "rendah"
    ukuranKota: str = "besar"
    
    # Data kendaraan
    mobil: int = 0
    bus: int = 0
    truk: int = 0
    motor: int = 0
    
    # Durasi observasi
    durasiMenit: int = 60
    waktuObservasi: str = ""


class CalculationResult(BaseModel):
    success: bool = True
    message: str
    data: Dict[str, Any]
